### PR TITLE
Slightly faster nexus

### PR DESCRIPTION
**Pull request by benjaminhwilliams**
Following a review of #279, I notice that in `format/nexus.py` we call `find_class` three times on the same handle every time we initialise either an `NXinstrument` or a `NXmxEntry`.  Obviously, this means performing the same iteration three times over all the children of that node.  I had a go at introducing a `find_classes` function that accepts multiple `nx_class` names and searches for them all in a single iteration.  Using @rjgildea's [performance test case](https://github.com/cctbx/dxtbx/issues/276#issue-784232041), it seems to save me an additional 10% of the already reduced run time.

With my changes:
```
$ dials.python time_format_class.py $(dials.data get -q vmxi_thaumatin)/image_15799_master.h5
1.5693 seconds (0.0157 seconds per image)
```

Without:
```
$ dials.python time_format_class.py $(dials.data get -q vmxi_thaumatin)/image_15799_master.h5
1.7475 seconds (0.0175 seconds per image)
```

### CODE REVIEW

File: format/nexus.py [hunk 1]
```diff
@@ -81,6 +83,45 @@ class NXValidationError(RuntimeError):
...
+def find_classes(node: NXNode, *nx_classes: Optional[str]) -> Tuple[List[h5py.Group]]:
+    """
+    Find instances of multiple NXclass types within the children of the current node.
+
+    Args:
+        node: The input h5py node (h5py.File or h5py.Group).
+        nx_classes: Names of NXclass types to search for.
+
+    Returns:
+        A list of matching nodes for each of the specified NXclass types.
```
Comment (rjgildea) [inline format/nexus.py:95]: It actually returns a tuple of lists

File: format/nexus.py [hunk 2]
```diff
@@ -81,6 +83,45 @@ class NXValidationError(RuntimeError):
...
+
+    Args:
+        node: The input h5py node (h5py.File or h5py.Group).
+        nx_class: Names of NXclass type to search for.
+
+    Returns:
+        The list of matching nodes for the specified NXclass type.
+    """
+    (instances,) = find_classes(node, nx_class)
+    return instances
```
Comment (rjgildea) [inline format/nexus.py:122]: Any advantage of your two-line version over doing this in a single line?
```suggestion
    return find_classes(node, nx_class)[0]
```

Comment (benjaminhwilliams) [reply format/nexus.py:122]: I think it's a tiny bit clearer for maintenance.  In the two-line way, it emphasises that the returned tuple from find_classes ought to have only one entry.  If it has more than one, then something is broken.  In the one-line version, such a breaking change would go unnoticed.

File: format/nexus.py (hunk 1 above)
Comment (benjaminhwilliams) [reply format/nexus.py:95]: Yes, but the return type is specified in the function signature, so I didn't think it necessary to labour the point here, since any API documentation will pick that up automatically.  I felt that, since functions that return multiple things usually return a tuple of those things, it was more important to describe succinctly what each of those things is.  If you prefer, I could reword it to 'A tuple containing a list of matching nodes for each of the specified NX_class types.'

Comment (benjaminhwilliams) [inline format/nexus.py:95]: In any case, this needs correcting.
```suggestion
        A list of matching nodes for each of the specified NX_class types.
```

File: format/nexus.py (hunk 2 above)
Comment (Anthchirp) [reply format/nexus.py:122]: I have to say I find the original confusing. It's only now that you spelled it out that I can see that this is what it's there for.
If you're really concerned about that case an `assert len()==1` makes the intent clear and also would give the opportunity to include a helpful error message for when it does crash out.

Comment (benjaminhwilliams) [reply format/nexus.py:122]: Hmm... with the benefit of a little time, I agree.  Thanks!  I think you're right and I don't think it needs the addition of an `assert`.  I think I was over-thinking it.
